### PR TITLE
WIP: AMI-1215 Integrate dynamic config

### PR DESCRIPTION
**Pull request by declanshanaghy**
PLEASEREVIEW @MatthaeusHarris 

Integrates dynamic config into sidecar.

TODO:
Finalize how CLOUD_STACK environment variable gets set. This will probably fall on supervisord config.
Retrieve eureka URL from the table rather than return an ENV variable.

### CODE REVIEW

File: bin/cli.js [hunk 1]
```diff
@@ -1,9 +1,11 @@
-#!/usr/bin/env node
+#!/usr/bin/env nodejs
```
Comment (MatthaeusHarris) [inline bin/cli.js]: Please leave this as node.  If you have nodejs installed (i.e. from the debian package), please symlink /usr/bin/nodejs to /usr/bin/node on your development machine.

File: lib/dynamic_config.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	metadata = {
+		"action": "initialize_config",
+		"status": "start",
+		"cloud_stack": this.cloud_stack
+	};
+	winston.info("Initializing dynamic config", metadata);
+	if ( this.cloud_stack ) {
+		regex = this.cloud_stack + '-[a-z]+-ConfigTable-[a-zA-Z0-9]+';
+		var db = new AWS.DynamoDB();
+		config = this;
```
Comment (MatthaeusHarris) [inline lib/dynamic_config.js]: I don't see where config is defined.  In javascript, using a variable without first defining it with var will create a global variable.

File: lib/dynamic_config.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+DynamicConfig.prototype.initialize = function(callback) {
+	metadata = {
+		"action": "initialize_config",
+		"status": "start",
+		"cloud_stack": this.cloud_stack
+	};
+	winston.info("Initializing dynamic config", metadata);
+	if ( this.cloud_stack ) {
+		regex = this.cloud_stack + '-[a-z]+-ConfigTable-[a-zA-Z0-9]+';
```
Comment (MatthaeusHarris) [inline lib/dynamic_config.js]: global?

File: lib/dynamic_config.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+function DynamicConfig(cloud_stack) {
+	this.cloud_stack = cloud_stack;
+	this.table_name = null;
+}
+
+module.exports = DynamicConfig;
+
+DynamicConfig.prototype.initialize = function(callback) {
+	metadata = {
```
Comment (MatthaeusHarris) [inline lib/dynamic_config.js]: global

File: lib/dynamic_config.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+				metadata = {
+					"action": "initialize_config",
+					"status": "list_tables",
+					"cloud_stack": config.cloud_stack,
+					"tables": data.TableNames,
+					"regex": regex
+				};
+				winston.info("Found tables", metadata);
+				for (var i in data.TableNames) {
+					table = data.TableNames[i];
```
Comment (MatthaeusHarris) [inline lib/dynamic_config.js]: global

File: lib/dynamic_config.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+					}
+				}
+			}
+			else {
+				callback(err, null);
+			}
+		});
+	}
+	else {
+		callback("cloud_stack not provided", null);
```
Comment (MatthaeusHarris) [inline lib/dynamic_config.js]: If you callback with new Error("cloud_stack not provided"), it creates a stack trace from the point where new Error() was called.  This can be useful in debugging.

File: lib/eureka-sidecar.js [hunk 7]
```diff
@@ -1,26 +1,30 @@
...
 			'timer': 5,
 			'retries': 5,
 			'retry_timeout': 5,
 			'eureka_version': 'v2',
-			'proto': 'http'
+			'proto': 'http',
+			'cloud_stack': null
 		};
-		for (var key in defaults) {
+		for (key in defaults) {
```
Comment (MatthaeusHarris) [inline lib/eureka-sidecar.js]: Why this change?  This makes key a global variable.

File: lib/eureka-sidecar.js [hunk 8]
```diff
@@ -1,26 +1,30 @@
...
 			options[key] = options[key] || defaults[key];
 		}
 
-		for (var key in required_options) {
+		for (key in required_options) {
```
Comment (MatthaeusHarris) [inline lib/eureka-sidecar.js]: Ditto

File: lib/eureka-sidecar.js [hunk 9]
```diff
@@ -29,42 +33,100 @@ module.exports = {
...
+			"action": "initialize",
+			"status": "start"
+		};
+		winston.info("Initializing eureka-sidecar", metadata);
 		this.options = this.validate_options(override_options);
-		console.log("Options: " );
-		console.log(this.options);
-		this.register();
-		this.setup_timer();	
+		this.retrieve_eureka_url();
```
Comment (MatthaeusHarris) [inline lib/eureka-sidecar.js]: Because this is a function that makes async calls, they won't be completed by the time the next line is executed.  This can cause race conditions.  Consider modifying retrieve_eureka_url to accept a callback, and wrapping the rest of the initialize function in that callback.

File: lib/eureka-sidecar.js [hunk 10]
```diff
@@ -29,42 +33,100 @@ module.exports = {
...
+							"action": "retrieve_eureka_url",
+							"status": "success",
+							"table": table,
+							"eureka_url": eureka_url
+						};
+						winston.info("Retrieved Eureka URL", metadata);
+					}
+				}.bind(this));
+			}
+		}.bind(this));
```
Comment (MatthaeusHarris) [inline lib/eureka-sidecar.js]: This is just a comment and not a critique:  instead of using .bind(this), another method I've seen used is to create a variable in the outer scope that references this (e.g. var that = this, and then using that.options.eureka_url instead of this.options.eureka_url).  It's purely a matter of taste.

File: lib/eureka-sidecar.js [hunk 11]
```diff
@@ -29,42 +33,100 @@ module.exports = {
...
+			"status": "start",
+			"retry_timeout": this.options.retry_timeout,
+			"attempt": this.attempt,
+			"retries": this.options.retries,
+			"eureka_url": this.options.eureka_url
+		};
+		winston.info("Registering with Eureka", metadata);
 		this.check_health(function(healthy) {
-			if (healthy) {
+			if (healthy && this.options.eureka_url) {
```
Comment (MatthaeusHarris) [inline lib/eureka-sidecar.js]: I think it's cleaner to set up your program flow such that you don't call eureka_register until this.options.eureka_url is populated.  See my comment in initialize.

File: lib/dynamic_config.js (hunk 2 above)
Comment (declanshanaghy) [inline lib/dynamic_config.js]: gotcha

Comment (declanshanaghy) [inline lib/dynamic_config.js]: Ahh that explains why i saw weirdness in the logging with the "metadata" variable.

File: lib/dynamic_config.js [hunk 12]
```diff
@@ -0,0 +1,66 @@
+var winston = require('winston');
+var AWS = require('aws-sdk');
+
+AWS.config.region = 'us-east-1';
```
Comment (instantlinux) [inline lib/dynamic_config.js]: spof - might need a way to override this in the future.
